File: wander/src/commonMain/kotlin/dev/ligature/wander/parser/Parser.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.wander.parser

import arrow.core.Either
import arrow.core.Either.Left
import arrow.core.Either.Right
import dev.ligature.gaze.*
import dev.ligature.wander.WanderError
import dev.ligature.wander.lexer.Token
import dev.ligature.wander.parser.Nibblers.scriptNib

data class ParsingError(override val message: String): WanderError

fun parse(script: List<Token>): Either<ParsingError, List<Element>> {
  val filteredInput = script.filter { token: Token ->
    token !is Token.Comment && token !is Token.Spaces && token !is Token.NewLine
  }.toList()
  val gaze = Gaze(filteredInput)
  return when(val res = gaze.attempt(scriptNib)) {
    null ->
      if (gaze.isComplete) {
        Right(listOf())
      } else {
        Left(ParsingError("No Match"))
      }
    else ->
      if (gaze.isComplete) {
        Right(res) // .filter(_.isDefined).map(_.get)))
      } else {
        Left(ParsingError("Not complete - ${gaze.peek()}"))
      }
  }
}

File: wander/shared/src/main/scala/dev/ligature/wander/new/Parser.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature.wander.`new`

import dev.ligature.gaze.{
  Gaze,
  Nibbler,
  optional,
  take,
  takeAll,
  takeCond,
  takeFirst,
  takeString,
  takeUntil,
  takeWhile,
  repeat
}
import dev.ligature.lig.LigNibblers

enum Element:
  case BooleanLiteral(value: Boolean)
  case Identifier(value: String)
  case Integer(value: Long)
  case StringLiteral(value: String)
  case Name(value: String)
  case Form(elements: Seq[Element])

def parse(input: Seq[Token]): Either[WanderError, Seq[Element]] = {
  val tokens = input.filter {
    _ match
      case Token.Space | Token.NewLine | Token.Comment(_) => false
      case _                                              => true
  }
  val gaze = Gaze(tokens)
  gaze.attempt(elementsNib) match {
    case None =>
      if gaze.isComplete then Right(List())
      else Left(WanderError("Error"))
    case Some(res) =>
      if gaze.isComplete then Right(res)
      else Left(WanderError("Error"))
  }
}

val integerElementNib: Nibbler[Token, Element] = gaze =>
  gaze.next() match
    case Some(Token.Integer(i)) => Some(List(Element.Integer(i)))
    case _                      => None

val stringElementNib: Nibbler[Token, Element] = gaze =>
  gaze.next() match
    case Some(Token.StringLiteral(s)) => Some(List(Element.StringLiteral(s)))
    case _                            => None

val booleanElementNib: Nibbler[Token, Element] = gaze =>
  gaze.next() match
    case Some(Token.BooleanLiteral(b)) => Some(List(Element.BooleanLiteral(b)))
    case _                             => None

val nameElementNib: Nibbler[Token, Element] = gaze =>
  gaze.next() match
    case Some(Token.Name(n)) => Some(List(Element.Name(n)))
    case _                   => None

val identifierElementNib: Nibbler[Token, Element] = gaze =>
  gaze.next() match
    case Some(Token.Identifier(i)) => Some(List(Element.Identifier(i)))
    case _                         => None

val elementsNib: Nibbler[Token, Element] =
  repeat(
    takeFirst(
      integerElementNib,
      stringElementNib,
      nameElementNib,
      booleanElementNib,
      identifierElementNib
    )
  )
